internal/config/cgroupmanager/systemdmanager.go

```diff
@@ -0,0 +1,183 @@
+// +build linux
+
+package cgroupmanager
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/cri-o/cri-o/utils"
+	"github.com/opencontainers/runc/libcontainer/cgroups/systemd"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const defaultSystemdParent = "system.slice"
+
+type Systemdv1Manager struct{}
+
+// Name returns the name of the cgroup manager (systemd)
+func (*Systemdv1Manager) Name() string {
+	return SystemdCgroupManager
+}
+
+// IsSystemd returns that it is a systemd cgroup manager
+func (*Systemdv1Manager) IsSystemd() bool {
+	return true
+}
+
+// GetContainerCgroupPath takes arguments sandbox parent cgroup and container ID and returns
+// the cgroup path for that containerID. If parentCgroup is empty, it
+// uses the default parent system.slice
+func (*Systemdv1Manager) GetContainerCgroupPath(sbParent, containerID string) string {
+	return getContainerCgroupPath(sbParent, containerID)
+}
+
+// GetSandboxCgroupPath takes the sandbox parent, and sandbox ID. It
+// returns the cgroup parent, cgroup path, and error.
+// it also checks there is enough memory in the given cgroup (4mb is needed for the runtime)
+func (*Systemdv1Manager) GetSandboxCgroupPath(sbParent, sbID string) (string, string, error) {
+	return getSandboxCgroupPathForSystemd(sbParent, sbID, cgroupMemorySubsystemMountPathV1, "memory.limit_in_bytes")
+}
+
+// MoveConmonToCgroup takes the container ID, cgroup parent, conmon's cgroup (from the config) and conmon's PID
+// It attempts to move conmon to the correct cgroup.
+func (*Systemdv1Manager) MoveConmonToCgroup(cid, cgroupParent, conmonCgroup string, pid int) string {
+	return moveConmonToSystemdCgroup(cid, cgroupParent, conmonCgroup, pid)
+}
+
+type Systemdv2Manager struct{}
+
+// Name returns the name of the cgroup manager (systemd)
+func (*Systemdv2Manager) Name() string {
+	return SystemdCgroupManager
+}
+
+// IsSystemd returns that it is a systemd cgroup manager
+func (*Systemdv2Manager) IsSystemd() bool {
+	return true
+}
+
+// GetContainerCgroupPath takes arguments sandbox parent cgroup and container ID and returns
+// the cgroup path for that containerID. If parentCgroup is empty, it
+// uses the default parent system.slice
+func (*Systemdv2Manager) GetContainerCgroupPath(sbParent, containerID string) string {
+	return getContainerCgroupPath(sbParent, containerID)
+}
+
+// systemd cgroups (v1 and v2) have no difference in creating the container cgroup path
```

Comment:
I would create `SystemdManager` with the common stuff and then derive it so you won't have to repeat the same methods for v1 and v2.

Reply:
ah good idea